File: src/list/reduce.ts
export default function reduce(
  input: number[],
  initial: number,
  fn: (a: number, b: number) => number
) {
  const output: number[] = [];
  if (input.length === 0) return output;
  output[0] = fn(initial, input[0]);
  for (let index = 1; index < input.length; index++)
    output[index] = fn(output[index - 1], input[index]);
  return output;
}

export function gcd(a: number, b: number): number {
  a = Math.abs(a);
  b = Math.abs(b);
  while (b) {
    const t = b;
    b = a % b;
    a = t;
  }
  return a;
}

export function lcm(a: number, b: number): number {
  return a && b ? Math.abs((a * b) / gcd(a, b)) : 0;
}

export function sum(a: number, b: number) {
  return a + b;
}

File: src/list/reduce.py
from collections.abc import Callable

def reduce(input: list[int], initial_value: int, fn: Callable[[int, int], int]):
  output: list[int] = []
  if len(input) == 0: return output
  output.append(fn(initial_value, input[0]))
  for index in range(1, len(input)):
    output.append(fn(output[index - 1], input[index]))
  return output

def gcd(a: int, bar: int):
  a = abs(a)
  b = abs(bar)
  while b != 0:
    t = b
    b = a % b
    a = t
  return a

def lcm(a: int, b: int):
  return 0 if a == 0 or b == 0 else abs((a * b) / gcd(a, b))
